pkg/apis/experimental/types.go

```diff
@@ -639,3 +639,36 @@ type ClusterAutoscalerList struct {
 
 	Items []ClusterAutoscaler `json:"items"`
 }
+
+// Selector is a label query over a set of API objects. MatchLabels and
+// MatchExpressions are ANDed. An empty Selector matches all objects. A
+// null Selector matches no objects.
+type Selector struct {
+	// MatchLabels is a list of labels.
+	MatchLabels map[string]string `json:"matchLabels,omitempty"`
+	// MatchExpressions is a list of requirement selectors.
+	MatchExpressions []SelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// SelectorRequirement is a selector that contains values, a key
+// and an operator that relates the key and values.
+type SelectorRequirement struct {
+	// Key is the label key that the selector applies to.
+	Key string `json:"key"`
+	// Operator represents a key's relationship
+	// to a set of values.
+	Operator SelectorOperator `json:"operator"`
+	// StringValues is a set of string values. If the operator is In or NotIn,
+	// the values set must be non-empty.
+	StringValues []string `json:"stringValues,omitempty"`
+}
+
+// SelectorOperator is the set of operators that can be used in
+// a selector requirement.
+type SelectorOperator string
+
+const (
+	SelectorOpIn     SelectorOperator = "In"
+	SelectorOpNotIn  SelectorOperator = "NotIn"
+	SelectorOpExists SelectorOperator = "Exists"
```

Comment:
Requirements:
* Indexability. For performance/scalability.
* Reverse indexability. For watch, we need to be able to find all outstanding queries that match a given label set as resources are modified.
* Understandability. It should be straightforward for a human to look at a selector and a label set and tell you whether the selector matches or not. Otherwise, users/operators will repeatedly shoot themselves in the foot.
* Easy proof of non-overlap. Important for controllers, load balancing, monitoring, and other cases. At least 1 key in common with different values is sufficient to prove non-overlap with the current semantics, which only permits contiguous subsets.